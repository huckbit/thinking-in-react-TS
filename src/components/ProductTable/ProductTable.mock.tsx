import type { Props as ProductTableProps } from "./ProductTable";

const mock: ProductTableProps = {
  products: [
    { category: "Fruits", price: "£1", stocked: true, name: "Apple" },
    { category: "Fruits", price: "£1", stocked: true, name: "Dragonfruit" },
    { category: "Fruits", price: "£2", stocked: false, name: "Passionfruit" },
  ],
  category: { category: "Fruits" },
  filterText: "apple",
  inStockOnly: true,
};

export default mock;
