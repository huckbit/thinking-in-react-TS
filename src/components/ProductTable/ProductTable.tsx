import React from "react";
import ProductRow, { Product } from "../ProductRow";
import ProductCategoryRow, { ProductCategoryRowProps } from "../ProductCategoryRow";

export interface Props {
  products: Product[];
  category?: ProductCategoryRowProps;
  filterText: string;
  inStockOnly?: boolean;
}

function ProductTable({ products, filterText, inStockOnly }: Props) {
  const rows: React.ReactNode[] = [];

  let lastCategory: String | null = null;

  products.forEach((product: Product, index: number) => {
    const { category } = product;
    if (product.name.toLowerCase().indexOf(filterText.toLowerCase()) === -1) {
      return;
    }
    if (inStockOnly && !product.stocked) {
      return;
    }
    if (category !== lastCategory) {
      rows.push(<ProductCategoryRow category={category} key={`category-${index}`} />);
    }
    rows.push(<ProductRow product={product} key={`product-${index}`} />);
    lastCategory = category;
  });

  return (
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Price</th>
        </tr>
      </thead>
      <tbody>{rows}</tbody>
    </table>
  );
}

export default ProductTable;
