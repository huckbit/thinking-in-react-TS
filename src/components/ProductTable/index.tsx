export { default } from "./ProductTable";
export type { Props as ProductTableProps } from "./ProductTable";
