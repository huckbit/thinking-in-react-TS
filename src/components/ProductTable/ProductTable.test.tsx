/// <reference types="@testing-library/jest-dom" />
import { render } from "@testing-library/react";
import ProductTable from "./ProductTable";
import mock from "./ProductTable.mock";

const container = render(
  <ProductTable
    products={mock.products}
    category={mock.category}
    filterText={mock.filterText}
    inStockOnly={mock.inStockOnly}
  />,
);

test("matches snapshot", () => {
  const { asFragment } = container;
  expect(asFragment()).toMatchSnapshot();
});

test("renders the correct number of products when filter is checked", () => {
  render(
    <ProductTable
      products={mock.products}
      category={mock.category}
      filterText={mock.filterText}
      inStockOnly={mock.inStockOnly}
    />,
  );
  expect(container.getAllByRole("row").length).toBe(3);
});
