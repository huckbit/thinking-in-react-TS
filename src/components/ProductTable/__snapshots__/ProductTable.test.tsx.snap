// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <table>
    <thead>
      <tr>
        <th>
          Name
        </th>
        <th>
          Price
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th
          colspan="2"
        >
          Fruits
        </th>
      </tr>
      <tr>
        <td>
          Apple
        </td>
        <td>
          Â£1
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
